C/services/core/configuration_manager.cpp

```diff
@@ -136,16 +137,261 @@ ConfigCategory ConfigurationManager::getCategoryItems(const string& categoryName
 
-	// Create category oject
-	ConfigCategory retVal = ConfigCategory(key->getString(), sItems);
+	// Create category object
+	ConfigCategory theVal(key->getString(), sItems);
 
 	// Set description
-	retVal.setDescription(description->getString());
+	theVal.setDescription(description->getString());
 
-	// Deallocate category items object
 	delete categoryItems;
 
-	return retVal;
+	return theVal;
+}
+
+/**
+ * Create or update a new category
+ *
+ * @param categoryName		The category name
+ * @param categoryDescription	The category description
+ * @param categoryItems		The category items
+ * @return			The ConfigCategory object
+ *				with "value" and "default"
+ *				of the new category added
+ *				or the merged configuration
+ *				of the updated confguration.
+ */
+
+ConfigCategory ConfigurationManager::createCategory(const std::string& categoryName,
+						    const std::string& categoryDescription,
+						    const std::string& categoryItems) const
+{
+	// Fill the ready to insert category object with input data
+	ConfigCategory preparedValue(categoryName, categoryItems);
+	preparedValue.setDescription(categoryDescription);
+
+	try
+	{
+		// Abort if items contain both value and default
+		preparedValue.checkDefaultValuesOnly();
+
+		// Add 'value' from 'default' for each item
+		preparedValue.setItemsValueFromDefault();
+	}
+	catch (ConfigMalformed* e)
+	{
+		delete e;
+		// TODO: add specific Exception
+		throw;
+	}
+	catch (ConfigValueFoundWithDefault* e)
+	{
+		delete e;
+		// TODO: add specific Exception
+		throw;
+	}
+	catch (...)
+	{
+		// TODO: add specific Exception
+		throw;
+	}
+
+	// Updated items JSON string
+        string updatedItems;
+
+	// SELECT * FROM foglamp.configuration WHERE key = categoryName
+	const Condition conditionKey(Equals);
+	Where *wKey = new Where("key", conditionKey, categoryName);
+	Query qKey(wKey);
+
+	// Query via storage client
+	ResultSet* result = m_storage->queryTable("configuration", qKey);
+	if (!result->rowCount())
+	{
+		// Prepare insert values for insertTable
+		InsertValues newCategory;
+		newCategory.push_back(InsertValue("key", categoryName));
+		newCategory.push_back(InsertValue("description", categoryDescription));
+		newCategory.push_back(InsertValue("value", preparedValue.itemsToJSON()));
+
+		if (!m_storage->insertTable("configuration", newCategory))
+		{
+			delete result;
+
+			// TODO: add specific Exception
+			throw;
+		}
+	}
+	else
+	{
+		// The category already exists: fetch data
+		ResultSet::RowIterator it = result->firstRow();
+		ResultSet::Row* row = *it;
+		if (!row)
+		{
+			delete result;
+
+			// TODO: add specific Exception
+			throw;
+		}
+
+		// Get current category items
+		ResultSet::ColumnValue* theItems = row->getColumn("value");
+		const Value* storedData = theItems->getJSON();
+
+		// Parse JSON input
+		Document doc;
+		// Parse the prepared input category with "value" and "default"
+		doc.Parse(preparedValue.itemsToJSON().c_str());
```

Comment:
Why are you converting a parsed config category to JSON in order to parse the JSON? Why not work from the original?

Reply:
The categoryItems in inout only has "default"
ConfigCategory preparedValue has "value" copied from "default"

So I use preparedValue.itemsToJSON()

We can change the interface and pass to this routine rapidjson::Value or ConfigCategory